### ISSUE TITLE
[Bug] 

### ISSUE THREAD

**Issue by Lelelo1** [xamarin/Xamarin.Forms#6476]
### Description

When defining a control  - and instantiating a new `DataTemplate` with it `new DataTemplate(() => control)` in code behind the content created by `createContent()` can only be added once to the page. If it was already there it won't create a new control object and add it. 

### Steps to Reproduce

1. Define a control
2. Create a Button with a clicked event handler just for testing.
3. In the clicked event handler instantiate a DataTemplate with the control
4. call `createContent` and add the content to the page

### Expected Behavior

The control should be added to the page each time the button is clicked.

### Actual Behavior

The control will only be added once. If the control was defined in xaml in the page - given to the `DataTemplate` with `x:name`- no content created by `createContent()` gets added.

### Basic Information

- Version with issue: 4.0.0.425677
- Last known good version: Unknown
- IDE: Visual Studio Community 2019 (Preview) for Mac Version 8.1 Preview (8.1 build 2697)
- Platform Target Frameworks: 
  - iOS:  Version: 12.10.0.153
  - Android: Untested
  - UWP:  Untested
- Android Support Library Version: Not used
- Nuget Packages: Xamarin.Essentials(1.1.0). Xamarin.Forms(4.0.0.425677)
- Affected Devices: iPhone 6s simulator. Other not yet tested.

### Reproduction Link

https://github.com/Lelelo1/DataTemplateReproduce

### More Info

Content created by a `DataTemplate` that is defined in xaml can successfully be added to the page. The Id in `Xamarin.Forms.Element` of this content is unique - while for the content from `DataTemplate` created in code behind - stays the same. It did not help to change the `_id` field with reflection prior to adding the content to the page.

**Comment by andreinitescu** [comment]
I think your sample works as expected, there's no bug.

The reason why `content1` is the same is because you're using a `DataTemplate` instance created based on your own `Func<object>` which just returns the **same** instance everytime you call `CreateContent()`:


            DataTemplate data = new DataTemplate(() => la); // use 'la', 'label' or 'l'. 
            var content1 = data.CreateContent();

DataTemplate just calls your func, that's all.

`content2` on the other hand is created using the `DataTemplate` defined in XAML:
```
<DataTemplate x:Key="dataTemplate" x:Name="dataTemplate">
       <Label Text="I am a label inside dataTemplate defined in xaml" />
</DataTemplate>
```
This one creates a new Label instance everytime you call `CreateContent()` because that's how what the code generated by the XAML compiler works like. It really behaves like it you would expect the "template" pattern to work: it generates a new content (view) based on the "template" you gave it (a Label). Under the hood, the generated IL calls the new c-tor on the root type you provided (Label) and returns the newly created instance.

**Comment by Lelelo1** [comment]
But isen't the point of`DataTemplate` to create a new instance from a given object? Would it also not be very inconsistent if the `DataTemplate` does it xaml- but not in code behind?

**Comment by andreinitescu** [comment]
> But isen't the point ofDataTemplate to create a new instance from a given object?

Correct, it is. But you are not doing that, you are returning same Label instance, `la`

**Comment by Lelelo1** [comment]
> > But isen't the point ofDataTemplate to create a new instance from a given object?
> 
> Correct, it is. But you are not doing that, you are returning same Label instance, `la`

It forces the developer, to move the initialising of the ui object to the inititialising of the `DataTemplate`. Meaning it is not possible to define a control in xaml and then pass it to a DataTemplate created in C# code behind or by any other means have them separated.

Right now the act of the `DataTemplate` seem to simply run a func where I am forced to create the ui into.

I don't think it is the same as create a new instance from a given object - but rather run a func where the ui is created.

**Comment by Lelelo1** [comment]
I am currently blocked somewhat related to this. I would like to attach an async event listener on `SizeChanged` for all content in `CollectionView` so I can set a `HeightRequest` that correspond to the height of the content.

I have made an extension method for `VisualElement` that returns a `Task<Rectangle>`. However it will only be attached to one object of the content created by the `DataTemplate`. I am forced to declare the UI with `new` keyword in C# inside the constructor of the `DataTemplate`.

**Comment by Lelelo1** [comment]
It's actually not true a new instance has to be created inside the func given to `DataTemplate` constructor. When given to a `CollectionView` the following works:

```
    Label testLabel = new Label() { Text = "Im a testLabel" };
    DataTemplate dataTemplate;
    public MyPage()
    {
        InitializeComponent();
        List<string> list = new List<string>() { "one", "two", "three", "four", "five", "six", "seven" };
        collectionView.ItemsSource = list;
        dataTemplate = new DataTemplate(() => testLabel); // <--
        collectionView.ItemTemplate = dataTemplate;
    }
```

![DataTemplateGIvenSameInstance](https://user-images.githubusercontent.com/42244983/60771335-cc2e0100-a0e6-11e9-86cc-801a8103c155.png)

But when using the `createContent()` method:

```
    void TestButton_Clicked(object sender, EventArgs eventArgs)
    {
        var content = (View) dataTemplate.CreateContent();
        rootStack.Children.Add(content);
    }

```

![TestButtonClicked](https://user-images.githubusercontent.com/42244983/60771347-fd0e3600-a0e6-11e9-8ed4-ec8b1cb57643.png)

...The `CollectionView` content goes missing. Also, clicking the Test button more times will not add additional testLabel content.
